libtrellis/tools/ecpmulti.cpp

```diff
@@ -0,0 +1,171 @@
+        help:
+        cerr << "Project Trellis - Open Source Tools for ECP5 FPGAs" << endl;
+        cerr << "ecpmulti: ECP5 multiboot bitstream assembler" << endl;
+        cerr << endl;
+        cerr << "Copyright (C) 2018 Jens Andersen <jens.andersen@gmail.com>" << endl;
+        cerr << endl;
+        cerr << options << endl;
+        return vm.count("help") ? 0 : 1;
+    }
+
+    if (vm.count("flashsize")) {
+        uint32_t flash_size_mbit = vm["flashsize"].as<uint32_t>();
+        flash_size_bytes = flash_size_mbit * 1024* 1024 / 8;
+        cout << "Using flashsize " << flash_size_mbit << "(" << flash_size_bytes << " bytes)" << endl;
+    }
+
+    auto inputs = vm.at("input").as<std::vector<std::string>>();
+    auto addresses = vm.at("address").as<std::vector<std::string>>();
+
+    if (inputs.size() != (addresses.size() + 1)) {
+        cerr << "Inputs " << inputs.size() << " offsets " << addresses.size() << endl;
+        cerr << "There must be one address specified for each input file" << endl;
+        return 1;
+    }
+
+    if (vm.count("db")) {
+        database_folder = vm["db"].as<string>();
+    }
+
+    try {
+        load_database(database_folder);
+    } catch (runtime_error &e) {
+        cerr << "Failed to load Trellis database: " << e.what() << endl;
+        return 1;
+    }
+
+    ofstream out_file(vm["output"].as<string>(), ofstream::trunc);
+    if (!out_file) {
+        cerr << "Failed to open output file" << endl;
+        return 1;
+    }
+    uint32_t idcode = 0;
+    uint32_t nextaddr = 0;
+    const map<string, string> bs_options = {{"multiboot", "yes"} };
+    for(uint32_t i = 0; i < inputs.size(); i++) {
+        string filename = inputs[i];
+        uint32_t address = nextaddr << 16;
+        cout << "Processing " << filename << "for address " << hex << address << endl;
+        ifstream bitfile(inputs[i], ios::binary);
+        Bitstream bs = Bitstream::read_bit(bitfile);
+        Chip c = bs.deserialise_chip();
+        if (idcode == 0)
+            idcode = c.info.idcode;
+
+        if (idcode != c.info.idcode) {
+            cerr << "All input files must be for the same model (" << idcode << " != " << c.info.idcode << ")" << endl;
+            return 1;
+        }
+
+        if (i < addresses.size()) {
+            string offset_str = addresses[i];
+            nextaddr = uint32_t(strtoul(offset_str.c_str(), nullptr, 0));
+            nextaddr = (nextaddr & 0x00FF0000) >> 16;
+            if (nextaddr == 0) {
+                cerr << "Invalid offset: " << offset_str << endl;
+                return 1;
+            }
+        } else nextaddr = 0;
+
+        auto tile_db = get_tile_bitdata(TileLocator{c.info.family, c.info.name, "EFB1_PICB1"});
+        auto tile = c.get_tiles_by_type("EFB1_PICB1");
+
+        WordSettingBits wsb = tile_db->get_data_for_setword("BOOTADDR");
+        int j=0;
+        cerr << "BOOTADDR has bits: " << wsb.bits[0].bits.size() << endl;
+        for(auto bg : wsb.bits) {
```

Comment:
A traditional for loop here would be neater than a range-based one and a counter.